connect/mirror/src/main/java/org/apache/kafka/connect/mirror/OffsetSyncStore.java

```diff
@@ -139,10 +171,103 @@ public void close() {
     protected void handleRecord(ConsumerRecord<byte[], byte[]> record) {
         OffsetSync offsetSync = OffsetSync.deserializeRecord(record);
         TopicPartition sourceTopicPartition = offsetSync.topicPartition();
-        offsetSyncs.put(sourceTopicPartition, offsetSync);
+        offsetSyncs.computeIfAbsent(sourceTopicPartition, ignored -> createInitialSyncs(offsetSync));
+        offsetSyncs.compute(sourceTopicPartition, (ignored, syncs) -> updateExistingSyncs(syncs, offsetSync));
+    }
+
+    private OffsetSync[] updateExistingSyncs(OffsetSync[] syncs, OffsetSync offsetSync) {
+        // Make a copy of the array before mutating it, so that readers do not see inconsistent data
+        // TODO: batch updates so that this copy can be performed less often for high-volume sync topics.
+        OffsetSync[] mutableSyncs = Arrays.copyOf(syncs, Long.SIZE);
+        updateSyncArray(mutableSyncs, offsetSync);
+        if (log.isTraceEnabled()) {
+            StringBuilder stateString = new StringBuilder();
+            stateString.append("[");
+            for (int i = 0; i < Long.SIZE; i++) {
+                if (i != 0) {
+                    stateString.append(",");
+                }
+                if (i == 0 || i == Long.SIZE - 1 || mutableSyncs[i] != mutableSyncs[i - 1]) {
+                    // Print only if the sync is interesting, a series of repeated syncs will appear as ,,,,,
+                    stateString.append(mutableSyncs[i].upstreamOffset());
+                    stateString.append(":");
+                    stateString.append(mutableSyncs[i].downstreamOffset());
+                }
+            }
+            stateString.append("]");
+            log.trace("New sync {} applied, new state is {}", offsetSync, stateString);
+        }
+        return mutableSyncs;
+    }
+
+    private OffsetSync[] createInitialSyncs(OffsetSync firstSync) {
+        OffsetSync[] syncs = new OffsetSync[Long.SIZE];
+        clearSyncArray(syncs, firstSync);
+        return syncs;
+    }
+
+    private void clearSyncArray(OffsetSync[] syncs, OffsetSync offsetSync) {
+        for (int i = 0; i < Long.SIZE; i++) {
+            syncs[i] = offsetSync;
+        }
+    }
+
+    private void updateSyncArray(OffsetSync[] syncs, OffsetSync offsetSync) {
+        long upstreamOffset = offsetSync.upstreamOffset();
+        // Old offsets are invalid, so overwrite them all.
+        if (!readToEnd || syncs[0].upstreamOffset() > upstreamOffset) {
```

Comment:
I needed a way to guarantee that if the task translates some offset, that it's at least as far along in the topic as the previous checkpoint. For two checkpoints emitted in the same task lifetime, the in-memory state in MirrorCheckpointTask is good enough to provide this guarantee.

If we restart the source task it will re-read the offset syncs topic and could potentially reconstruct a different in-memory state, both in the MirrorCheckpointTask and the OffsetSyncStore. By discarding all of the earlier syncs, we ensure that the earliest sync we have in-memory has a higher offset than every previous offset sync that the previous lifetime of the OffsetSyncStore could have had.

Consider the following situation:
1. An upstream consumer group is close to the replication flow, everything is in steady state. The offset syncs array contains syncs A (oldest), B, and C (newest)
2. A checkpoint is emitted which uses a fine grained offset sync C.
3. The upstream consumer group stops consuming.
4. MirrorSourceTask emits offset sync D, and the MirrorCheckpointTask discards offset sync C because it is of similar precision.
5. The OffsetSyncStore translates the consumer group relative to offset sync B, but the MirrorCheckpointTask drops this checkpoint because it remembers that it previously translated relative to C.
6. The MirrorCheckpointTask restarts, and in-memory checkpoints are cleared and offsets are restored.
7. The MirrorCheckpointTask now tries to perform a checkpoint.

With the current `!readToEnd` condition, only offset sync D is in-memory, and translation is skipped because the consumer group is behind the latest restart.
If we remove the `!readToEnd` condition, the OffsetSyncStore translates the consumer group relative to offset sync B, and the MirrorCheckpointTask emits the checkpoint because it does not remember the previous checkpoint emitted.
Thus it emitted a checkpoint computed by C, and then B, which was non-monotonic.

I mentioned in a comment on KAFKA-14666 that we would have to re-read the checkpoints topic for arbitrarily old translation, and this is why. It can be done, it just represented a new complexity that I wanted to avoid.

Reply:
Hmmm... it feels like we're sacrificing a fair amount here in the name of correctness.

The first naive thought that comes to mind is that we could list the offsets for each to-be-synced group on the target cluster before doing the first sync for it, and refuse to translate backwards.

Of course, this doesn't handle cases where 1) the upstream offsets went backwards for some reason, or 2) when the downstream consumer group has been active and committed later offsets between now and when the last `MirrorCheckpointTask` instance was running.

I don't know if 2) should really be a concern or not (if you've committed offsets in that group, it's debatable whether we should be overwriting those). But I think 1) is worth taking into account, and I'm not sure how we can address it. Re-reading the offsets sync topic on startup and checking to see a decrease in upstream offsets between two consecutive syncs doesn't take compaction into account.

I'd love it if we could leverage some of this information that's in the offsets sync topic on startup so that we can narrow the likelihood of inactive or slow-moving upstream consumer groups failing to be synced, but I agree that if the scenario you outlined is possible (and without having thought too hard about it, it does seem that way), we can't sacrifice correctness by ignoring it.

Reply:
One idea to this:
What if the checkpointing started using the source offset mechanism of Connect to keep track of the last offset-sync used for a specific consumer group?
MirrorCheckpointTask could start emitting a source offset record like
(group, topic, partition) -> (offsetOfLastUsedOffsetSync)

This would allow
1. The OffsetSyncStore to read compaction-eligible offset syncs at startup, since it will be able to detect that (going with the example from @gharris1727) it cannot use offset-sync B, as the last saved offset of the offset-sync topic was C (which has a higher offset than B).
2. To rewind consumer group offsets if the source topic was recreated (offset sync record offsets will keep increasing regardless of the reset in the upstream offset)

I understand that in complexity, this is pretty much the same as reading back checkpoints, but in terms of implementation, it should be simpler (?). One drawback is that the deletion of the offset-syncs topic messes up the offset translation logic - but I guess with KIP-875 (or with the followup of that KIP containing the "reset" functionality) that won't be a problem, as the checkpoint source offsets can be cleared to restore a clean state.

Reply:
@urbandan I hadn't considered using the source offsets, and this is worth discussing more in a follow-up, as I think the current implementation with its limitations is all that we can afford for the upcoming release.

Compared to re-reading the checkpoints topic, the advantages of adding checkpoint source offsets are:
1. We can use the source offsets consumer in the framework
2. The framework will compact the data for us.

The disadvantages are:
1. Deletion of the offset syncs loses monotonicity guarantees
2. In non-EOS mode we lose monotonicity guarantees (source offsets arrive after checkpoints are written)
3. New load to the global source offsets topic that was not there before, hosting duplicate data.

Reply:
I agree that it could be a followup, but I would argue that the monotonicity guarantee is not required.
1. I think it is not a regression, as up until the latest fixes in checkpointing there were no such guarantees (e.g. a task restart couldused to be able to trigger a checkpoint emission using old offset syncs)
2. The best checkpointing can do is at-least-once, with or without the EOS feature. With that, comes possible duplication, so MM2 checkpoint users always need to prepare for de-duplication.

With these in mind, I think it would be a fair trade-off to drop the monotonicity guarantee. Additionally, this tradeoff will only affect the checkpoints, but not the sync.group.offsets feature, as that one checks the committed offsets in the target cluster.

I'm just bringing this up because even if this change doesn't implement the source offset usage in checkpointing, we can still simplify the change if we drop the monotonicity requirement now.

Reply:
@urbandan Before https://issues.apache.org/jira/browse/KAFKA-13659 and https://issues.apache.org/jira/browse/KAFKA-12468 this logic did not have monotonicity guarantees, but that does not mean that behavior was acceptable to users. These tickets were opened and voted on by multiple users before we took action.

If we eliminate the in-memory deduplication and the forced read-to-end in this PR, we will re-introduce behavior that we have already fixed on trunk. I do not think that is acceptable.

Reply:
@gharris1727 I had the wrong impression about the behavior pre-KAFKA-13659, I thought that it was capable of checkpointing older source offsets after restart, but I was wrong.

Regardless of that, what I'm trying to say is
1. Monotonicity is an optimization - it tries to minimize re-processing after a failover. (Please correct me if I'm wrong, but I don't really see any other user stories behind it.)
2. Being able to checkpoint old offsets after a restart is a feature, and probably a good one. If cluster restarts/rebalances are frequent enough, and some consumers are lagging behind consitently, they might never get their checkpoints translated, ever.
3. Based on our discussion so far, it seems that monotonicity will not allow us to implement "old offset checkpointing after restart", which is not ideal. Worst case of breaking monotonicity is sub-optimal failover and extra re-processing. Worst case for not translating old offsets after restart is never checkpointing a group. Then, with the default consumer config (auto.offset.reset=latest), failover results in data loss.

Reply:
@urbandan 

> Monotonicity is an optimization - it tries to minimize re-processing after a failover. (Please correct me if I'm wrong, but I don't really see any other user stories behind it.)

Yes this is correct, but I think it misses some nuance. The overall feature (offset translation, checkpoints, syncing offsets to the consumer group) is an optimization: users could just replicate their data, and start reading the replica from the beginning. And relevant to your earlier simplicity argument: it would be simpler to not offer offset translation, but would require users to re-deliver more data.

In the same way, monotonicity is an optimization to not rewind checkpoints when we have already translated better ones: We could choose to not offer monotonicity for the simplicity but (with this specific in-memory store implementation) that means that if you are N offsets behind the replication flow, your offset could fall backward by another N offsets, doubling the data re-delivery. 

I would also consider the user-story of someone monitoring their cluster, and noticing that despite consumers and replication moving forward, the checkpoints topic is moving _backwards_. Though no error has occurred, the result of the offset translation is getting worse. To someone unfamiliar with the internal workings of this algorithm, it looks like we're just generating arbitrary offsets. Additionally, causes a significant difference in semantics between using the raw checkpoints topic and synced consumer offsets, which I think is more likely to lead to confusion.

> Being able to checkpoint old offsets after a restart is a feature, and probably a good one. If cluster restarts/rebalances are frequent enough, and some consumers are lagging behind consistently, they might never get their checkpoints translated, ever.

I completely agree. I think that offset translation _must_ be extended to offer translation of very old offsets, and the solution in this PR is just part of that overall solution. The limitation you pointed out is a very important one, as it's the core limitation that this PR is trying to solve. Before this PR, consumers lagging behind the latest offset sync couldn't be translated. After this PR, consumers lagging behind the latest checkpoint task restart can't be translated. I think that is a strict improvement, and even if this PR doesn't address the issue completely, it is worth merging to get a release out-the-door.

> Based on our discussion so far, it seems that monotonicity will not allow us to implement "old offset checkpointing after restart", which is not ideal. Worst case of breaking monotonicity is sub-optimal failover and extra re-processing. Worst case for not translating old offsets after restart is never checkpointing a group - then, with the default consumer config (auto.offset.reset=latest), failover results in data loss.

I don't think the auto.offset.reset=latest situation is a fair criticism of this PR, as that can happen with or without this change. And actually, it's more likely to happen without this change, as the translation window is smaller.

You are welcome to help in designing and implementing the full back-translation algorithm and help get it landed in 3.5.1/3.6. This PR is a tactical change to get us over the line to 3.5.0, and is necessarily going to leave some things on the table. I'm glad that you are interested in MM2 offset translation, as I think full back-translation may be more than I can take on alone. Are you interested in exploring these ideas yourself and opening a follow-up PR?

Reply:
@gharris1727 I agree with almost all of what you said, but I would like to focus on the last point:

> > Based on our discussion so far, it seems that monotonicity will not allow us to implement "old offset checkpointing after restart", which is not ideal. Worst case of breaking monotonicity is sub-optimal failover and extra re-processing. Worst case for not translating old offsets after restart is never checkpointing a group - then, with the default consumer config (auto.offset.reset=latest), failover results in data loss.
> 
> I don't think the auto.offset.reset=latest situation is a fair criticism of this PR, as that can happen with or without this change. And actually, it's more likely to happen without this change, as the translation window is smaller.

Sorry, I'm not trying to critique this PR with these points, but to critique the monotonicity requirement. My understanding is that this PR would be the one making the monotonicity guarantee "official", and I'm worried that if it is indeed a tradeoff between monotonicity and old offset translation, we already made a choice, making the latter impossible. I am interested in the future of offset translation and checkpointing, and I can think of possible solutions, but (again, based on my current understanding) those will be limited if we have to keep monotonicity as a guarantee.

As an extra note, I'm familiar with the situation behind KAFKA-13659, and in that case, the symptom triggering the investigation was not non-monotonic checkpoints, but the weird "feature" of copying source offsets as-is when there was no available offset-sync, which caused negative lag in the target cluster. The checkpoints were monotonic, but they didn't make sense.

Reply:
I think the most convincing point in favor of monotonicity is that violations of it will seem like a bug to users and will almost certainly lead to Jira tickets being logged in the future asking us to fix that behavior.

I think it's important to keep in mind that exactly-once offset syncing is impossible no matter what, and in addition, any consumer applications that cannot tolerate data loss should already have `auto.offset.reset` set to `earliest`, regardless of whether we make this change or not.

Reply:
I understand the points about monotonicity, and if noticed by a user, it will be a problem, but I don't really expect users noticing it:
1. The current API to access checkpoints (in MirrorClient) only returns the latest checkpoint, and it is usually used for failover. Users would need to do advanced/custom monitoring to keep scraping checkpoints and notice non-monotonic checkpoints, which seems highly unlikely to me. Even if it is the case, we could just document that monotonicity is not guaranteed.
2. The auto offset sync feature checks the committed offsets in the target cluster, and does not rewind, so it is already protected in this sense.

I understand that my points are not convincing (enough) to make an impact on this change - as @gharris1727 suggested, I will try to think of a new solution for offset translation.